toolchain/lex/lex.cpp

```diff
@@ -94,21 +97,38 @@ class [[clang::internal_linkage]] Lexer {
 
   auto current_line() -> LineIndex { return LineIndex(line_index_); }
 
-  auto current_line_info() -> TokenizedBuffer::LineInfo* {
+  auto current_line_info() -> LineInfo* {
     return &buffer_.line_infos_[line_index_];
   }
 
   auto next_line() -> LineIndex { return LineIndex(line_index_ + 1); }
 
-  auto next_line_info() -> TokenizedBuffer::LineInfo* {
+  auto next_line_info() -> LineInfo* {
     CARBON_CHECK(line_index_ + 1 <
                  static_cast<ssize_t>(buffer_.line_infos_.size()));
     return &buffer_.line_infos_[line_index_ + 1];
   }
 
-  auto NoteWhitespace() -> void {
-    buffer_.token_infos_.back().has_trailing_space = true;
-  }
+  // Note when the lexer has encountered whitespace, and the next lexed token
+  // should reflect that it was preceded by some amount of whitespace.
+  auto NoteWhitespace() -> void;
+
+  // Add a lexed token to the tokenized buffer, and reset any token-specific
+  // state tracked in the lexer for the next token.
+  auto AddLexedToken(TokenInfo info) -> TokenIndex;
+
+  // An overload for lexing a token.Builds the correctly encoded token info,
+  // adds it to the tokenized buffer and returns the token index.
+  //
+  // This overload is for tokens without a payload.
+  auto LexToken(TokenKind kind, int32_t byte_offset) -> TokenIndex;
+
+  // An overload for lexing a token.Builds the correctly encoded token info,
+  // adds it to the tokenized buffer and returns the token index.
+  //
+  // This overload is for tokens with a payload.
+  auto LexToken(TokenKind kind, int token_payload, int32_t byte_offset)
```

Comment:
```suggestion
  auto LexToken(TokenKind kind, int token_payload, int32_t byte_offset)
```

It seems like it would be better to put `token_payload` last. When overload sets have different numbers of arguments, the shorter parameter list is usually a prefix of the longer one, as if the overload set was a single function with defaulted parameters. It seems especially important for the parameter order to be unsurprising in this case, since the caller generally won't get a type error if they get it wrong.

Reply:
The downside is that the payload is tightly coupled with the kind, while the byte offset is much less interesting and useful to think about separately.

On the balance, I find the current structure more readable, if not perfect.

But if this is a big issue for folks, we don't need these to be overloads -- would it help to have a name to distinguish them?

Reply:
I think a separate name will help somewhat, but it won't fully solve the problem unless the name is so different that the reader doesn't think of them as related (which seems undesirable).

I see your point about the problem with putting the payload last. To me it seems like the lesser evil, but it's your call.

Reply:
Did the rename, I think I prefer sticking with that for now.